### DISCUSSION TITLE
Force close a connection by the server

### DISCUSSION THREAD

**Discussion by izaaz** [dotnet/aspnetcore#29507] [Q&A] [answered]
We have a need to force close connections from the server every once in a while. The way we do this is by setting the connection close header when needed. Something like this (in this example we send connection close to all requests)

```csharp
        public void UseConnectionCloseMiddleware(IApplicationBuilder app)
        {
            app.Use(
                async (context, next) =>
                {
                    // setting Connection Header to 'close' for closeConnectionPercentage% of the connections
                    context.Response.Headers.Add("Connection", "close");
                    await next.Invoke();
                });
        }
```

This works fine when we run the service with kestrel. The moment IIS comes into the picture, it looks like IIS removes this header and does not send it back to the caller. Is there a way to stop IIS from doing this? I tried InProcess and OutOfProcess hosting models and both don't support this functionality.

**Comment by Tratcher** [comment, marked as answer]
Did you confirm the client is using HTTP/1.1 for these requests? That trick won't work for HTTP/2 (except on Http.Sys).

**Reply by izaaz** [reply]
Thanks for your response Chris. I'm making calls using postman and confirmed that postman is using HTTP/1.1. 

I took the basic weather forecast project, added the above connection close middleware and tested it as follows
1. Run the app as a standalone server without IIS locally - server returns the header correctly and I see the response header in postman
2. Run the app locally with IIS Express - server returns the header to IIS but IIS does not return this header back to postman.
3. Run the app as a webapp on Azure - same as the previous point.

**Reply by Tratcher** [reply]
This worked for me. I used:
- Fiddler
- Kestrel and IIS Express in-proc 
- .NET 5.
- The Empty template

```
                endpoints.MapGet("/", async context =>
                {
                    context.Response.Headers.Add("Connection", "close");
                    await context.Response.WriteAsync("Hello World!");
                });
```

The `Connection: close` header is returned by both servers.

Note this isn't expected to work in an azure web app. Connection is a hop-by-hop header and azure uses load balancers. That means you end up closing the connection between the app and the load balancer but not the connection between the load balancer and the client. You'll have a similar problem with IIS out-of-proc.

**Reply by izaaz** [reply]
Thanks Chris. This ended up being a postman problem and not an issue with aspnetcore. I updated to the latest version of postman and I see connection close header back.
